File: proto/order_test.go
package order_test

import (
	"testing"

	"buf.build/go/protovalidate"
	pb "bulldog_alpha/proto/order"
)

func TestOrderValidation(t *testing.T) {
	v, err := protovalidate.New()
	if err != nil {
		t.Fatalf("failed to initialize validator: %v", err)
	}

	tests := []struct {
		name    string
		req     *pb.OrderRequest
		wantErr bool
	}{
		{
			name: "valid order",
			req: &pb.OrderRequest{
				OrderId:       "123",
				Symbol:        "AAPL",
				Price:         150.0,
				Quantity:      10,
				Side:          pb.OrderSide_BUY,
				Type:          pb.OrderType_LIMIT,
				CorrelationId: "corr-1",
			},
			wantErr: false,
		},
		{
			name: "invalid symbol - lowercase",
			req: &pb.OrderRequest{
				OrderId:       "123",
				Symbol:        "aapl",
				Price:         150.0,
				Quantity:      10,
				Side:          pb.OrderSide_BUY,
				Type:          pb.OrderType_LIMIT,
				CorrelationId: "corr-2",
			},
			wantErr: true,
		},
		{
			name: "invalid symbol - empty",
			req: &pb.OrderRequest{
				OrderId:       "123",
				Symbol:        "",
				Price:         150.0,
				Quantity:      10,
				Side:          pb.OrderSide_BUY,
				Type:          pb.OrderType_LIMIT,
				CorrelationId: "corr-3",
			},
			wantErr: true,
		},
		{
			name: "invalid price - negative",
			req: &pb.OrderRequest{
				OrderId:       "123",
				Symbol:        "AAPL",
				Price:         -10.0,
				Quantity:      10,
				Side:          pb.OrderSide_BUY,
				Type:          pb.OrderType_LIMIT,
				CorrelationId: "corr-4",
			},
			wantErr: true,
		},
		{
			name: "invalid quantity - zero",
			req: &pb.OrderRequest{
				OrderId:       "123",
				Symbol:        "AAPL",
				Price:         150.0,
				Quantity:      0,
				Side:          pb.OrderSide_BUY,
				Type:          pb.OrderType_LIMIT,
				CorrelationId: "corr-5",
			},
			wantErr: true,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			err := v.Validate(tc.req)
			if tc.wantErr {
				if err == nil {
					t.Error("expected validation error, got none")
				} else {
					t.Logf("got expected error: %v", err)
				}
			} else {
				if err != nil {
					t.Errorf("expected no validation error, got: %v", err)
				}
			}
		})
	}
}

File: proto/order_test.py
import unittest
from proto import order_pb2
from proto import market_data_pb2

class TestOrderProto(unittest.TestCase):
    def test_order_fields(self):
        req = order_pb2.OrderRequest()
        req.order_id = "123"
        req.symbol = "AAPL"
        req.price = 150.0
        req.quantity = 10.0
        req.side = order_pb2.OrderSide.BUY
        req.type = order_pb2.OrderType.LIMIT
        req.correlation_id = "corr-123"
        
        self.assertEqual(req.order_id, "123")
        self.assertEqual(req.symbol, "AAPL")
        self.assertEqual(req.price, 150.0)
        self.assertEqual(req.quantity, 10.0)
        self.assertEqual(req.side, order_pb2.OrderSide.BUY)
        self.assertEqual(req.type, order_pb2.OrderType.LIMIT)
        self.assertEqual(req.correlation_id, "corr-123")

    def test_market_data_fields(self):
        tick = market_data_pb2.EquityTick()
        tick.symbol = "AAPL"
        tick.price = 150.0
        tick.size = 100.0
        tick.timestamp = 1600000000000
        tick.correlation_id = "corr-456"
        
        self.assertEqual(tick.symbol, "AAPL")
        self.assertEqual(tick.price, 150.0)
        self.assertEqual(tick.size, 100.0)
        self.assertEqual(tick.timestamp, 1600000000000)
        self.assertEqual(tick.correlation_id, "corr-456")

if __name__ == "__main__":
    unittest.main()
